File: app/src/main/java/ru/magflayer/spectrum/presentation/common/android/navigation/holder/BaseRouterHolder.java
package ru.magflayer.spectrum.presentation.common.android.navigation.holder;

import ru.terrakok.cicerone.BaseRouter;
import ru.terrakok.cicerone.Cicerone;
import ru.terrakok.cicerone.Navigator;
import ru.terrakok.cicerone.NavigatorHolder;

public abstract class BaseRouterHolder<T extends BaseRouter> {

    private Cicerone<T> cicerone;

    BaseRouterHolder(T router) {
        cicerone = Cicerone.create(router);
    }

    public T getRouter() {
        return cicerone.getRouter();
    }

    public NavigatorHolder getNavigatorHolder() {
        return cicerone.getNavigatorHolder();
    }

    public void setNavigator(final Navigator navigator) {
        getNavigatorHolder().setNavigator(navigator);
    }

    public void removeNavigator() {
        getNavigatorHolder().removeNavigator();
    }

}

File: app/src/main/java/ru/magflayer/spectrum/presentation/common/android/navigation/holder/BaseRouterHolder.kt
package ru.magflayer.spectrum.presentation.common.android.navigation.holder

import com.github.terrakok.cicerone.BaseRouter
import com.github.terrakok.cicerone.Cicerone
import com.github.terrakok.cicerone.Navigator
import com.github.terrakok.cicerone.NavigatorHolder

abstract class BaseRouterHolder<T : BaseRouter> internal constructor(router: T) {

    private val cicerone: Cicerone<T> = Cicerone.create(router)

    val router: T
        get() = cicerone.router

    private val navigatorHolder: NavigatorHolder
        get() = cicerone.getNavigatorHolder()

    fun setNavigator(navigator: Navigator) {
        navigatorHolder.setNavigator(navigator)
    }

    fun removeNavigator() {
        navigatorHolder.removeNavigator()
    }

}
